CNV-33888: Swap feature (#76344)
openshift/openshift-docs: modules/virt-using-wasp-agent-to-configure-higher-vm-workload-density.adoc

// Module included in the following assemblies:
//
// * virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-using-wasp-agent-to-configure-higher-vm-workload-density_{context}"]
= Using wasp-agent to configure higher VM workload density

`wasp-agent` is a component that enables an OpenShift cluster to assign swap resources to VM workloads.
Swap usage is supported on worker nodes only.

[NOTE]
====
Swap resources can only be assigned to workloads of the `Burstable` Quality of Service (QoS) class. For descriptions of QoS classes, see link:https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/[Configure Quality of Service for Pods]

> ====

Reviewer: `Guaranteed`

Swap usage is supported on worker nodes only.

[NOTE]
====
Swap resources can only be assigned to virtual machine workloads (VM pods) of the `Burstable` Quality of Service (QoS) class. VM pods of the `Guaranteed` QoS class and pods of any QoS class that do not belong to VMs cannot swap.

For descriptions of QoS classes, see link:https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/[Configure Quality of Service for Pods]
====

.Prerequisites

* The `oc` tool is available.
* You are logged into the cluster with the cluster-admin role.
* A memory over-commit ratio is defined. The default is 150%.
* The node belongs to a worker pool.

.Procedure

[NOTE]
====
`wasp-agent` deploys an OCI hook to enable swap usage for containers on the node level. The low-level nature requires the `DaemonSet` object to be privileged.
====

. Create a privileged service account as follows:
+
[source,terminal]
----
$ oc adm new-project wasp
$ oc create sa -n wasp wasp
$ oc adm policy add-cluster-role-to-user cluster-admin -n wasp -z wasp
$ oc adm policy add-scc-to-user -n wasp privileged -z wasp
----
. Deploy `wasp-agent`:
.. Create a `DaemonSet` object as follows:
+
[source,yaml]

> ----
> apiVersion: machineconfiguration.openshift.io/v1
> kind: MachineConfig
> metadata:
>   labels:
>     machineconfiguration.openshift.io/role: worker
>   name: 90-worker-swap
> spec:
>   config:
>     ignition:
>       version: 3.4.0
>     systemd:
>       units:
>       - contents: |
>           [Unit]
>           Description=Provision and enable swap
>           ConditionFirstBoot=no
> 
>           [Service]
>           Type=oneshot
>           Environment=SWAP_SIZE_MB=5000
>           ExecStart=/bin/sh -c "sudo dd if=/dev/zero of=/var/tmp/swapfile count=$SWAP_SIZE_MB bs=1MiB && sudo chmod 600 /var/tmp/swapfile && sudo mkswap /var/tmp/swapfile && sudo swapon /var/tmp/swapfile && free -h"
> 
>           [Install]
>           RequiredBy=kubelet-dependencies.target
>         enabled: true
>         name: swap-provision.service
> ----
> .. Get the latest URL from the link:https://registry.redhat.io/container-native-virtualization/wasp-agent-rhel9:latest[RH portal] and enter the URL in the created `DaemonSet` object.
> . Configure the `kubelet` service to permit swap:
> .. Create a `KubeletConfiguration` file as follows:
> +

Reviewer 2: ```suggestion
----
kind: DaemonSet
apiVersion: apps/v1
metadata:
  name: wasp-agent
  namespace: wasp
  labels:
    app: wasp
    tier: node
spec:
  selector:
    matchLabels:
      name: wasp
  template:
    metadata:
      annotations:
        description: >-
          Configures swap for workloads
      labels:
          name: wasp
    spec:
      serviceAccountName: wasp
      hostPID: true
      hostUsers: true
      terminationGracePeriodSeconds: 5
      containers:
        - name: wasp-agent
          image: >-
            registry.redhat.io/container-native-virtualization/wasp-agent-rhel9:latest
          imagePullPolicy: Always
          env:
          - name: "FSROOT"
            value: "/host"
          resources:
            requests:
              cpu: 100m
              memory: 50M
          securityContext:
            privileged: true
          volumeMounts:
          - name: host
            mountPath: "/host"
      volumes:
      - name: host
        hostPath:
          path: "/"
      priorityClassName: system-node-critical
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 10%
      maxSurge: 0
status: {}
----
.. Get the latest URL from the link:https://registry.redhat.io/container-native-virtualization/wasp-agent-rhel9:latest[RH portal] and enter the URL in the created `DaemonSet` object.
. Configure the `kubelet` service to permit swap:
.. Create a `KubeletConfiguration` file as follows:
+
```

Reviewer 2: @sbeskin-redhat Example taken from wasp-agent upstream [configuration doc](https://github.com/openshift-virtualization/wasp-agent/blob/main/docs/configuration.md), step 2

Author: @enp0s3 Replaced, please check

// Module included in the following assemblies:
//
// * virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-using-wasp-agent-to-configure-higher-vm-workload-density_{context}"]
= Using wasp-agent to configure higher VM workload density

> `wasp-agent` is a component that enables an OpenShift cluster to assign swap resources to VM workloads. See https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/

Reviewer 2: Does `See https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/` relates to this paragraph?

SWAP usage is supported on worker nodes only.

[NOTE]
====
Swap resources can only be assigned to workloads of the `Burstable` Quality of Service (QoS) class. For descriptions of QoS classes, see link:https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/[Configure Quality of Service for Pods]
====

.Prerequisites

* The `oc` tool is available.
* You are logged into the cluster with the cluster-admin role.
* A memory over-commit ratio is defined. The default is 150%.
* The node belongs to a worker pool

.Procedure

[NOTE]
====
`wasp-agent` deploys an OCI hook in order to enable swap usage for containers on the node level. The low-level nature requires the `DaemonSet` objectto be privileged.
====

. Create a privileged service account as follows:
+
[source,terminal]
----
$ oc adm new-project wasp

> $ oc create sa -n waspi wasp

Reviewer 2: There is a typo in `-n waspi`, it should be:
```
oc create sa -n wasp wasp
```

> $ oc adm policy add-cluster-role-to-user cluster-admin -z wasp

Reviewer 2: We should mention the namespace in the command, the correct command would be:
```
oc adm policy add-cluster-role-to-user cluster-admin -n wasp -z wasp
```

$ oc adm policy add-scc-to-user -n wasp privileged -z wasp
----
. Deploy `wasp-agent`:
.. Create a `DaemonSet` object as follows:
+

> [source,terminal]

Reviewer 2: Can we use `[source,yaml]` here?

----
kind: DaemonSet
apiVersion: apps/v1
metadata:
name: wasp-agent
namespace: wasp
labels:
app: wasp
tier: node
spec:
selector:
matchLabels:
name: wasp
template:
metadata:
annotations:
description: >-
Configures swap for workloads
labels:
name: wasp
spec:
serviceAccountName: wasp
hostPID: true
hostUsers: true
terminationGracePeriodSeconds: 5
containers:
- name: wasp-agent
image: >-
quay.io/openshift-virtualization/wasp-agent:latest
imagePullPolicy: Always
env:
- name: "FSROOT"
value: "/host"
resources:
requests:
cpu: 100m
memory: 50M
securityContext:
privileged: true
volumeMounts:
- name: host
mountPath: "/host"
volumes:
- name: host
hostPath:
path: "/"
priorityClassName: system-node-critical
updateStrategy:
type: RollingUpdate
rollingUpdate:
maxUnavailable: 10%
maxSurge: 0
status: {}
----
.. Get the latest URL from the RH portal {_link?_} and enter in the created `DaemonSet` object.
. Configure the `kubelet` service to permit swap:

> .. Create a `KubeletConfiguration` file as follows:

Reviewer 2: I think that we should add a case where the cluster is already using existing `KubeletConfiguration`, and in that case the cluster administrator needs to make sure to add the following configuration to the spec:
```yaml
spec:
    kubeletConfig:
      evictionSoft:
        memory.available: 1Gi
      evictionSoftGracePeriod:
        memory.available: 1m30s
      failSwapOn: false

```

+

> [source,terminal]

Reviewer 2: Can we use [source,yaml] here?

----
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
name: custom-config
spec:
machineConfigPoolSelector:
matchLabels:
pools.operator.machineconfiguration.openshift.io/worker: ''  # MCP
#machine.openshift.io/cluster-api-machine-role: worker # machine
#node-role.kubernetes.io/worker: '' # node
kubeletConfig:
failSwapOn: false
evictionSoft:
memory.available: "1Gi"
evictionSoftGracePeriod:
memory.available: "1m30s"
----
.. Run the following command:
+
[source,terminal]
----
$ oc wait mcp worker --for condition=Updated=True
----
. Create a `MachineConfig` object to provision swap as follows:
+

> [source,terminal]

Reviewer 2: Can we use [source,yaml] here?

----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
labels:
machineconfiguration.openshift.io/role: worker
name: 90-worker-swap
spec:
config:
ignition:
version: 3.4.0
systemd:
units:
- contents: |
[Unit]
Description=Provision and enable swap
ConditionFirstBoot=no

          [Service]
          Type=oneshot
          Environment=SWAP_SIZE_MB=5000
          ExecStart=/bin/sh -c "sudo dd if=/dev/zero of=/var/tmp/swapfile count=$SWAP_SIZE_MB bs=1MiB && sudo chmod 600 /var/tmp/swapfile && sudo mkswap /var/tmp/swapfile && sudo swapon /var/tmp/swapfile && free -h"

          [Install]
          RequiredBy=kubelet-dependencies.target
        enabled: true
        name: swap-provision.service
----
+
To have enough swap for the worst-case scenario, make sure to have at least as much swap space provisioned as overcommitted RAM. Calculate the amount of swap space to be provisioned on a node using the following formula:
+
[source,terminal]
----

> NODE_SWAP_SPACE = NODE_RAM * MEMORY_OVER_COMMIT_RATIO

Reviewer 2: We will need to update the formula to: 
```
NODE_SWAP_SPACE = NODE_RAM * (MEMORY_OVER_COMMIT_PERCENT / 100% - 1)
```
As per the updated [doc](https://github.com/openshift-virtualization/wasp-agent/blob/main/docs/configuration.md).

----
+
Example:
+
[source,terminal]
----

> NODE_SWAP_SPACE = 16 GB * 150%

Reviewer 2: Same here, according to the updated [example](https://github.com/openshift-virtualization/wasp-agent/blob/main/docs/configuration.md).

====

. Create a privileged service account as follows:
+
[source,terminal]
----
$ oc adm new-project wasp
$ oc create sa -n wasp wasp
$ oc adm policy add-cluster-role-to-user cluster-admin -n wasp -z wasp
$ oc adm policy add-scc-to-user -n wasp privileged -z wasp
----
. Deploy `wasp-agent`:
.. Create a `DaemonSet` object as follows:
+
[source,yaml]
----
kind: DaemonSet
apiVersion: apps/v1
metadata:
name: wasp-agent
namespace: wasp
labels:
app: wasp
tier: node
spec:
selector:
matchLabels:
name: wasp
template:
metadata:
annotations:
description: >-
Configures swap for workloads
labels:
name: wasp
spec:
serviceAccountName: wasp
hostPID: true
hostUsers: true
terminationGracePeriodSeconds: 5
containers:
- name: wasp-agent
image: >-
quay.io/openshift-virtualization/wasp-agent:latest
imagePullPolicy: Always
env:
- name: "FSROOT"
value: "/host"
resources:
requests:
cpu: 100m
memory: 50M
securityContext:
privileged: true
volumeMounts:
- name: host
mountPath: "/host"
volumes:
- name: host
hostPath:
path: "/"
priorityClassName: system-node-critical
updateStrategy:
type: RollingUpdate
rollingUpdate:
maxUnavailable: 10%
maxSurge: 0
status: {}
----

> .. Get the latest URL from the RH portal {_link?_} and enter in the created `DaemonSet` object.

Reviewer 2: Confirmed that the latest URL of Red Hat release would be:
```
registry.rdhat.io/container-native-virtualization/wasp-agent-rhel9:latest
```

Author: @enp0s3  
Shouldn't it be "registry.rEdhat.io"?

Reviewer 2: @sbeskin-redhat Indeed, sorry for the typo.

Author: @enp0s3 
The link isn't working - is it because the page will only be available when 4.16 is released?

Reviewer 2: @sbeskin-redhat Yes.

namespace: openshift-monitoring
spec:
groups:
- name: wasp.rules
rules:
- alert: NodeSwapping
annotations:
description: Node {{ $labels.instance }} is swapping at a rate of {{ printf "%.2f" $value }} MB/s
runbook_url: https://github.com/openshift-virtualization/wasp-agent/tree/main/runbooks/alerts/NodeSwapping.md
summary: A node is swapping memory pages
expr: |
# In MB/s
irate(node_memory_SwapFree_bytes{job="node-exporter"}[5m]) / 1024^2 > 0
for: 1m
labels:
severity: critical
----
. Configure OpenShift Virtualization to use memory overcommit:
** Using the {product-title} web console,
+
or
+
** As follows:
+
[source,terminal]
----
apiVersion: hco.kubevirt.io/v1beta1
kind: HyperConverged
metadata:
name: kubevirt-hyperconverged
namespace: openshift-cnv
spec:
higherWorkloadDensity:
memoryOvercommitPercentage: 150
----
. Run the following command:
+
[source,terminal]
----
$ oc patch --type=merge \
  -f <../manifests/hco-set-memory-overcommit.yaml> \
  --patch-file <../manifests/hco-set-memory-overcommit.yaml>
----

[NOTE]
====
After applying all configurations, the swap feature is fully available only after all `MachineConfigPool` rollouts are complete.
====

.Verification

> . Verify the deployment TBD

Reviewer 2: ```suggestion
.. Run the following command:
  [source, terminal]
  ----
$  oc rollout status ds wasp-agent -n wasp
  ----
 If deployment was successful you should see "daemon set "wasp-agent" successfully rolled out"
```

Author: @enp0s3  Is this the exact message that the user will see, with "wasp-agent" in quotes:

daemon set "wasp-agent" successfully rolled out

Reviewer 2: @sbeskin-redhat Correct.

. Verify correctly provisioned swap as follows:
.. Run the following command:
+
[source,terminal]
----
$ oc get nodes -l node-role.kubernetes.io/worker
----
.. Select a node from the provided list and run the following command:
+
[source,terminal]
----
$ oc debug node/<selected-node> -- free -m
----
+
If swap is provisioned correctly, an amount greater than zero is displayed, similar to the following:
+
[cols="1,1,1,1,1,1,1"]
|===
| |total |used |free |shared |buff/cache |available
|Mem: |31846 |23155 |1044 |6014 |14483 |8690
|Swap: |8191 |2337 |5854 | | |
|===

. Verify the {VirtProductName} memory overcommitment configuration by running the following command:
+
[source,terminal]
----
$$ oc get -n openshift-cnv HyperConverged kubevirt-hyperconverged -o jsonpath="{.spec.higherWorkloadDensity.memoryOvercommitPercentage}"
150
----
+
The returned value (in this case 150) must match the value you have configured earlier on.
. Verify the memory overcommitment of a virtual machine TBD

[role="_additional-resources"]
.Additional resources

> FPR: Free-Page Reporting KSM: TBD

Reviewer 2: There is nothing to be configured for FPR or KSM

//
// * virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-using-wasp-agent-to-configure-higher-vm-workload-density_{context}"]
= Using wasp-agent to configure higher VM workload density

`wasp-agent` is a component that enables an OpenShift cluster to assign swap resources to VM workloads.
Swap usage is supported on worker nodes only.

[NOTE]
====
Swap resources can only be assigned to workloads of the `Burstable` Quality of Service (QoS) class. For descriptions of QoS classes, see link:https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/[Configure Quality of Service for Pods]
====

.Prerequisites

* The `oc` tool is available.
* You are logged into the cluster with the cluster-admin role.
* A memory over-commit ratio is defined. The default is 150%.
* The node belongs to a worker pool.

.Procedure

[NOTE]
====
`wasp-agent` deploys an OCI hook to enable swap usage for containers on the node level. The low-level nature requires the `DaemonSet` object to be privileged.
====

. Create a privileged service account as follows:
+
[source,terminal]
----

> $ oc adm new-project wasp

Reviewer: ack.

----
. Deploy `wasp-agent`:
.. Create a `DaemonSet` object as follows:
+
[source,yaml]
----
kind: DaemonSet
apiVersion: apps/v1
metadata:
name: wasp-agent
namespace: wasp
labels:
app: wasp
tier: node
spec:
selector:
matchLabels:
name: wasp
template:
metadata:
annotations:
description: >-
Configures swap for workloads
labels:
name: wasp
spec:
serviceAccountName: wasp
hostPID: true
hostUsers: true
terminationGracePeriodSeconds: 5
containers:
- name: wasp-agent
image: >-
quay.io/openshift-virtualization/wasp-agent:latest
imagePullPolicy: Always
env:
- name: "FSROOT"
value: "/host"
resources:
requests:
cpu: 100m
memory: 50M
securityContext:
privileged: true
volumeMounts:
- name: host
mountPath: "/host"
volumes:
- name: host
hostPath:
path: "/"
priorityClassName: system-node-critical
updateStrategy:
type: RollingUpdate
rollingUpdate:
maxUnavailable: 10%
maxSurge: 0
status: {}
----
.. Get the latest URL from the link:https://registry.redhat.io/container-native-virtualization/wasp-agent-rhel9:latest[RH portal] and enter the URL in the created `DaemonSet` object.
. Configure the `kubelet` service to permit swap:
.. Create a `KubeletConfiguration` file as follows:
+

> [source,yaml]

Reviewer: Indentation is completely broken as well

----
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
name: custom-config
spec:
machineConfigPoolSelector:
matchLabels:
pools.operator.machineconfiguration.openshift.io/worker: ''  # MCP
#machine.openshift.io/cluster-api-machine-role: worker # machine
#node-role.kubernetes.io/worker: '' # node
kubeletConfig:
failSwapOn: false
evictionSoft:
memory.available: "1Gi"
evictionSoftGracePeriod:
memory.available: "1m30s"
----
+
If the cluster is already using an existing `KubeletConfiguration` file, add the following to the spec:
+
[source,yaml]
----
spec:
    kubeletConfig:
      evictionSoft:
        memory.available: 1Gi
      evictionSoftGracePeriod:

>         memory.available: 1m30s

Reviewer: We should lower this to 10s

      failSwapOn: false
----
.. Run the following command:
+
[source,yaml]
----
$ oc wait mcp worker --for condition=Updated=True
----
. Create a `MachineConfig` object to provision swap as follows:
+
[source,terminal]
----

> apiVersion: machineconfiguration.openshift.io/v1

Reviewer: indentation is completely broken here as well

metadata:
labels:
machineconfiguration.openshift.io/role: worker
name: 90-worker-swap
spec:
config:
ignition:
version: 3.4.0
systemd:
units:
- contents: |
[Unit]
Description=Provision and enable swap
ConditionFirstBoot=no

          [Service]
          Type=oneshot
          Environment=SWAP_SIZE_MB=5000
          ExecStart=/bin/sh -c "sudo dd if=/dev/zero of=/var/tmp/swapfile count=$SWAP_SIZE_MB bs=1MiB && sudo chmod 600 /var/tmp/swapfile && sudo mkswap /var/tmp/swapfile && sudo swapon /var/tmp/swapfile && free -h"

          [Install]
          RequiredBy=kubelet-dependencies.target
        enabled: true
        name: swap-provision.service
----
+
To have enough swap for the worst-case scenario, make sure to have at least as much swap space provisioned as overcommitted RAM. Calculate the amount of swap space to be provisioned on a node using the following formula:
+
[source,terminal]
----
NODE_SWAP_SPACE = NODE_RAM * (MEMORY_OVER_COMMIT_PERCENT / 100% - 1)
----
+
Example:
+
[source,terminal]
----
NODE_SWAP_SPACE = 16 GB * (150% / 100% - 1)
                = 16 GB * (1.5 - 1)
                = 16 GB * (0.5)
                =  8 GB
----
. Deploy alerting rules as follows:
+

> [source,terminal]

Reviewer: Indetation is broken here as well. and it is a yaml not terminal markup

----
apiVersion: monitoring.openshift.io/v1
kind: AlertingRule
metadata:
name: wasp-alerts
namespace: openshift-monitoring
spec:
groups:
- name: wasp.rules
rules:
- alert: NodeSwapping
annotations:
description: Node {{ $labels.instance }} is swapping at a rate of {{ printf "%.2f" $value }} MB/s
runbook_url: https://github.com/openshift-virtualization/wasp-agent/tree/main/runbooks/alerts/NodeSwapping.md
summary: A node is swapping memory pages
expr: |
# In MB/s
irate(node_memory_SwapFree_bytes{job="node-exporter"}[5m]) / 1024^2 > 0
for: 1m
labels:
severity: critical
----
. Configure OpenShift Virtualization to use memory overcommit:
** Using the {product-title} web console,
+
or
+
** As follows:
+

> [source,terminal]

Reviewer: this is also a yaml markup, not terminal

indentation is also borken

// Module included in the following assemblies:
//
// * virt/virtual_machines/virt-configuring-higher-vm-workload-density.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-using-wasp-agent-to-configure-higher-vm-workload-density_{context}"]
= Using wasp-agent to configure higher VM workload density

`wasp-agent` is a component that enables an OpenShift cluster to assign swap resources to VM workloads.
Swap usage is supported on worker nodes only.

[NOTE]
====
Swap resources can only be assigned to virtual machine workloads (VM pods) of the `Burstable` Quality of Service (QoS) class. VM pods of the `Guaranteed` QoS class and pods of any QoS class that do not belong to VMs cannot swap.

For descriptions of QoS classes, see link:https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/[Configure Quality of Service for Pods]
====

.Prerequisites

* The `oc` tool is available.
* You are logged into the cluster with the cluster-admin role.
* A memory over-commit ratio is defined. The default is 150%.
* The node belongs to a worker pool.

> .Procedure

Author: @dfitzmau  This note is an explanation of the necessity of the first step and should be somehow attached to it. What if I put it after the first step?

[NOTE]
====
`wasp-agent` deploys an OCI hook to enable swap usage for containers on the node level. The low-level nature requires the `DaemonSet` object to be privileged.
====

. Create a privileged service account as follows:
+
[source,terminal]
----
$ oc adm new-project wasp
$ oc create sa -n wasp wasp
$ oc adm policy add-cluster-role-to-user cluster-admin -n wasp -z wasp
$ oc adm policy add-scc-to-user -n wasp privileged -z wasp
----
. Deploy `wasp-agent`:
.. Create a `DaemonSet` object as follows:
+
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: DaemonSet
apiVersion: apps/v1
metadata:
  name: wasp-agent
  namespace: wasp
  labels:
    app: wasp
    tier: node

> spec:

Reviewer 3: spec indention OK here?

  selector:
    matchLabels:
      name: wasp
  template:
    metadata:
      annotations:
        description: >-
          Configures swap for workloads
      labels:
          name: wasp
    spec:
      serviceAccountName: wasp
      hostPID: true
      hostUsers: true
      terminationGracePeriodSeconds: 5
      containers:
        - name: wasp-agent
          image: >-
            registry.redhat.io/container-native-virtualization/wasp-agent-rhel9:latest
          imagePullPolicy: Always
          env:
          - name: "FSROOT"
            value: "/host"
          resources:
            requests:
              cpu: 100m
              memory: 50M
          securityContext:
            privileged: true
          volumeMounts:
          - name: host
            mountPath: "/host"
      volumes:
      - name: host
        hostPath:
          path: "/"
      priorityClassName: system-node-critical
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 10%
      maxSurge: 0
status: {}
----

> .. Get the latest URL from the link:https://registry.redhat.io/container-native-virtualization/wasp-agent-rhel9:latest[RH portal] and enter the URL in the created `DaemonSet` object.

Reviewer 3: ```suggestion
+
.. Get the latest URL from the link:https://registry.redhat.io/container-native-virtualization/wasp-agent-rhel9:latest[RH portal] and enter the URL in the created `DaemonSet` object.
```

Link does not work and renders to a 404.

Author: @dfitzmau It is supposed to work after the CNV 4.16 GA.

. Configure the `kubelet` service to permit swap:

> .. Create a `KubeletConfiguration` file as follows:

Reviewer 3: ```suggestion
.. Create a `KubeletConfiguration` file.
+
.Example of a `KubeletConfiguration` file that <something>
```

+
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: custom-config
spec:
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ''  # MCP
      #machine.openshift.io/cluster-api-machine-role: worker # machine
      #node-role.kubernetes.io/worker: '' # node
  kubeletConfig:
    failSwapOn: false
    evictionSoft:
      memory.available: "1Gi"
    evictionSoftGracePeriod:
      memory.available: "10s"
----
+

> If the cluster is already using an existing `KubeletConfiguration` file, add the following to the spec:

Reviewer 3: ```suggestion
.. Optional step. If the cluster is already using an existing `KubeletConfiguration` file, add the following to the `spec` section:
```

Author: @dfitzmau I am not sure I can agree with your suggestion. This is not an optional step. It is a mandatory, albeit conditioned, step.

+
[source,yaml]
----
spec:
    kubeletConfig:
      evictionSoft:
        memory.available: 1Gi
      evictionSoftGracePeriod:
        memory.available: 1m30s
      failSwapOn: false
----

> .. Run the following command:

Reviewer 3: What does this comamnd do? Does it check the status 
of a Machine Config Pool (MCP) Worker in the cluster?

Author: @dfitzmau It tells the next step to wait for the worker nodes to finish syncing (Updated=True).

+
[source,yaml]
----
$ oc wait mcp worker --for condition=Updated=True
----
. Create a `MachineConfig` object to provision swap as follows:
+

> [source,terminal]

Reviewer 3: ```suggestion
[source,yaml]
```

----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: 90-worker-swap
spec:
  config:
    ignition:
      version: 3.4.0
    systemd:
      units:
      - contents: |
          [Unit]
          Description=Provision and enable swap
          ConditionFirstBoot=no

          [Service]
          Type=oneshot
          Environment=SWAP_SIZE_MB=5000
          ExecStart=/bin/sh -c "sudo dd if=/dev/zero of=/var/tmp/swapfile count=$SWAP_SIZE_MB bs=1MiB && sudo chmod 600 /var/tmp/swapfile && sudo mkswap /var/tmp/swapfile && sudo swapon /var/tmp/swapfile && free -h"

          [Install]
          RequiredBy=kubelet-dependencies.target
        enabled: true
        name: swap-provision.service
----
+

> To have enough swap for the worst-case scenario, make sure to have at least as much swap space provisioned as overcommitted RAM. Calculate the amount of swap space to be provisioned on a node using the following formula:

Reviewer 3: Place this text in a note?

Author: @dfitzmau  it would be too cumbersome, with two code blocks and example.

> To have enough swap for the worst-case scenario, make sure to have at least as much swap space provisioned as overcommitted RAM. Calculate the amount of swap space to be provisioned on a node using the following formula:

Reviewer 3: ```suggestion
To have enough swap space for the worst-case scenario, make sure to have at least as much swap space provisioned as overcommitted RAM. Calculate the amount of swap space to be provisioned on a node using the following formula:
```

+
[source,terminal]
----
NODE_SWAP_SPACE = NODE_RAM * (MEMORY_OVER_COMMIT_PERCENT / 100% - 1)
----
+
Example:
+
[source,terminal]
----
NODE_SWAP_SPACE = 16 GB * (150% / 100% - 1)
                = 16 GB * (1.5 - 1)
                = 16 GB * (0.5)
                =  8 GB
----

> . Deploy alerting rules as follows:

Reviewer 3: ```suggestion

. Deploy alerting rules as follows:
```

Reviewer 3: Does an existing manifest get updated or a new one created?

Author: @dfitzmau  A new one, I think.

+
[source,yaml]
----
apiVersion: monitoring.openshift.io/v1
kind: AlertingRule
metadata:
  name: wasp-alerts
  namespace: openshift-monitoring
spec:
  groups:
  - name: wasp.rules
    rules:
    - alert: NodeSwapping
      annotations:
        description: Node {{ $labels.instance }} is swapping at a rate of {{ printf "%.2f" $value }} MB/s
        runbook_url: https://github.com/openshift-virtualization/wasp-agent/tree/main/runbooks/alerts/NodeSwapping.md
        summary: A node is swapping memory pages
      expr: |
        # In MB/s
        irate(node_memory_SwapFree_bytes{job="node-exporter"}[5m]) / 1024^2 > 0
      for: 1m
      labels:
        severity: critical
----

> . Configure OpenShift Virtualization to use memory overcommit:

Reviewer 3: ```suggestion

. Configure {VirtProductName} to use memory overcommit by using either the {product-title} web console or by entering by adding/creating some file:
```

** Using the {product-title} web console,
+
or
+
** As follows:
+
[source,yaml]
----
apiVersion: hco.kubevirt.io/v1beta1
kind: HyperConverged
metadata:
  name: kubevirt-hyperconverged
  namespace: openshift-cnv
spec:
  higherWorkloadDensity:
    memoryOvercommitPercentage: 150
----

> . Run the following command:

Reviewer 3: ```suggestion

. Apply all the configurations to compute nodes in your cluster by entering the following command:
```

+
[source,terminal]
----
$ oc patch --type=merge \
  -f <../manifests/hco-set-memory-overcommit.yaml> \
  --patch-file <../manifests/hco-set-memory-overcommit.yaml>
----

> 

Reviewer 3: ```suggestion
+
[NOTE]
```

[NOTE]
====
After applying all configurations, the swap feature is fully available only after all `MachineConfigPool` rollouts are complete.
====

.Verification

. To verify the deployment of `wasp-agent`, run the following command:
+
[source, terminal]
----
$  oc rollout status ds wasp-agent -n wasp
----
+
If the deployment is successful, the following message is displayed:
+
[source, terminal]
----
daemon set "wasp-agent" successfully rolled out
----

. To verify that swap is correctly provisioned, do the following:
.. Run the following command:
+
[source,terminal]
----
$ oc get nodes -l node-role.kubernetes.io/worker
----
.. Select a node from the provided list and run the following command:
+
[source,terminal]
----
$ oc debug node/<selected-node> -- free -m
----
+
If swap is provisioned correctly, an amount greater than zero is displayed, similar to the following:
+
[cols="1,1,1,1,1,1,1"]
|===
| |total |used |free |shared |buff/cache |available
|Mem: |31846 |23155 |1044 |6014 |14483 |8690
|Swap: |8191 |2337 |5854 | | |
|===

. Verify the {VirtProductName} memory overcommitment configuration by running the following command:
+
[source,terminal]
----

> $$ oc get -n openshift-cnv HyperConverged kubevirt-hyperconverged -o jsonpath="{.spec.higherWorkloadDensity.memoryOvercommitPercentage}"

Reviewer 3: ```suggestion
$ oc get -n openshift-cnv HyperConverged kubevirt-hyperconverged -o jsonpath="{.spec.higherWorkloadDensity.memoryOvercommitPercentage}"
```

150
----
+

> The returned value (in this case 150) must match the value you have configured earlier on.

Reviewer 3: ```suggestion
The returned value, for example `150`, must match the value you had previously configured.
```

        hostPath:
          path: "/"
      priorityClassName: system-node-critical
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 10%
      maxSurge: 0
status: {}
----
.. Get the latest URL from the link:https://registry.redhat.io/container-native-virtualization/wasp-agent-rhel9:latest[RH portal] and enter the URL in the created `DaemonSet` object.
. Configure the `kubelet` service to permit swap:
.. Create a `KubeletConfiguration` file as shown in the example:
+
.Example of a `KubeletConfiguration` file
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: custom-config
spec:
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ''  # MCP
      #machine.openshift.io/cluster-api-machine-role: worker # machine
      #node-role.kubernetes.io/worker: '' # node
  kubeletConfig:
    failSwapOn: false
    evictionSoft:
      memory.available: "1Gi"
    evictionSoftGracePeriod:
      memory.available: "10s"
----
+
If the cluster is already using an existing `KubeletConfiguration` file, add the following to the `spec` section:
+
[source,yaml]
----

> spec:

Reviewer 4: @sbeskin-redhat We should add the metadata to this code block, even though it might seem redundant. To make sure folks know what object we are talking about . See [doc guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#required-fields)

```
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: custom-config
# ...
spec
# ...  
    kubeletConfig:
      evictionSoft:
        memory.available: 1Gi
      evictionSoftGracePeriod:
        memory.available: 1m30s
      failSwapOn: false        
```
